miasm2/analysis/mem.py

```diff
@@ -0,0 +1,1259 @@
+
+    @value.setter
+    def value(self, s):
+        """Get the string value from memory"""
+        if self._enc == "ansi":
+            set_str = set_str_ansi
+        elif self._enc == "utf16":
+            set_str = set_str_utf16
+        else:
+            raise NotImplementedError("Only 'ansi' and 'utf16' are implemented")
+        set_str(self._vm, self.get_addr(), s)
+
+    def get_size(self):
+        """This get_size implementation is quite unsafe: it reads the string
+        underneath to determine the size, it may therefore read a lot of memory
+        and provoke mem faults (analogous to strlen).
+        """
+        val = self.value
+        if self._enc == "ansi":
+            return len(val) + 1
+        elif self._enc == "utf16":
+            # FIXME: real encoding...
+            return len(val) * 2 + 2
+        else:
+            raise NotImplementedError("Only 'ansi' and 'utf16' are implemented")
+
+    def raw(self):
+        raw = self._vm.get_mem(self.get_addr(), self.get_size())
+        return raw
+
+    def __repr__(self):
+        return "%r(%s): %r" % (self.__class__, self._enc, self.value)
+
+
+class MemArray(MemStruct):
+    """An unsized array of type @field_type (a MemField subclass instance).
+    This class has no static or dynamic size.
+
+    It can be indexed for setting and getting elements, example:
+
+        array = MemArray(vm, addr, Num("I"))
+        array[2] = 5
+        array[4:8] = [0, 1, 2, 3]
+        print array[20]
+
+    If the @field_type is a Ptr, deref_get(index) and deref_set(index) can be
+    used to dereference a field at a given index in the array.
+
+    mem_array_type can be used to generate a type that includes the field_type.
+    Such a generated type can be instanciated with only vm and addr, as are
+    other MemStructs.
+    """
+    _field_type = None
+
+    def __init__(self, vm, addr=None, field_type=None):
+        if self._field_type is None:
+            self._field_type = field_type
+        if self._field_type is None:
+            raise NotImplementedError(
+                "Provide field_type to instanciate this class, "
+                "or generate a subclass with mem_array_type.")
+        super(MemArray, self).__init__(vm, addr)
+
+    @property
+    def field_type(self):
+        """Return the MemField subclass instance that represents the type of
+        this MemArray items.
+        """
+        return self._field_type
+
+    def _normalize_idx(self, idx):
+        # Noop for this type
+        return idx
+
+    def _normalize_slice(self, slice_):
+        start = slice_.start if slice_.start is not None else 0
+        stop = slice_.stop if slice_.stop is not None else self.get_size()
+        step = slice_.step if slice_.step is not None else 1
+        return slice(start, stop, step)
+
+    def _check_bounds(self, idx):
+        idx = self._normalize_idx(idx)
+        if not isinstance(idx, int) and not isinstance(idx, long):
```

Comment:
In Python, you can use `isinstance(idx, (int, long))`, or, for this specific case, `isinstance(idx, numbers.Number)` which also work for `float` and `Decimal`